### PR TITLE
Deterministic index selection for filter rule with no hybrid scan

### PR DESCRIPTION
**Pull request by apoorvedave1**
### What is the context for this pull request?
Deterministic index selection for FilterIndexRule with hybrid scan disabled.

 - **Dependent PRs**: https://github.com/microsoft/hyperspace/pull/386

### What changes were proposed in this pull request?
This PR changes the filter index rule selection from picking the `head` of the eligible indexes to picking the index which has
- the minimum size in terms of index file sizes
- lexicographically smallest index name (if there are multiple indexes with minimum size)

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

unit tests

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRanker.scala
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRanker.scala:62]: Could you use `sourceFilesSizeInBytes` in IndexLogEntry?

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRanker.scala:62]: i think that is for source data, not index data

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRanker.scala:62]: pl. let me know if i am wrong

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:116]: This function is for hybrid scan.

File: src/main/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRanker.scala
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRanker.scala:62]: ok.. then can we just add `lazy val indexFilesSizeInBytes` instead of tag?
since I think we can use this condition for join index ranker.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala
Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:116]: sorry I don't understand why we can't use this function. Could you please explain why we need to limit its use to hybrid scan?

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:116]: For gold standard tests, there are no files present on disk so to test this functionality, I will need to create index files. Just trying to avoid that case

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:116]: This change is for "non" hybrid scan, but if hybrid scan is not enabled, this function is not called.

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:116]: oh i see what you are saying, from your previous comment i thought "setTagValue" was created specifically for hybrid scan.

File: src/test/scala/com/microsoft/hyperspace/index/HyperspaceSuite.scala
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/HyperspaceSuite.scala:21]: nit: --

File: src/test/scala/com/microsoft/hyperspace/index/HyperspaceSuite.scala
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/HyperspaceSuite.scala:27]: nit: --

File: src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala:106]: Is the change in this file required?

File: src/test/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRankerTest.scala
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rankers/FilterIndexRankerTest.scala:58]: comments don't seem to match up?

File: src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala:32]: You can change this to `defaultFilenames` and use it as a default parameter.

File: src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala:106]: yes without this system path could be relative

File: src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleSuite.scala:32]: thanks @imback82

**Review by imback82 — APPROVED** [review]
LGTM (pending @sezruby's approval), thanks @apoorvedave1!

**Comment by sezruby** [comment]
LGTM Thanks!

**Pull request merged**
